core/domain/stats_jobs_one_off.py

```diff
@@ -231,21 +231,21 @@ def _apply_state_name_changes(cls, prev_stats_dict, change_list):
             dict. A dict representation of an ExplorationStatsModel
                 with updated state_stats_mapping and version.
         """
+        exp_versions_diff = exp_domain.ExplorationVersionsDiff(change_list)
+
         # Handling state additions, deletions and renames.
-        for change_dict in change_list:
-            if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
-
-                prev_stats_dict['state_stats_mapping'][change_dict[
-                    'state_name']] = (stats_domain.StateStats
-                                      .create_default())
-            elif change_dict['cmd'] == exp_domain.CMD_DELETE_STATE:
-                prev_stats_dict['state_stats_mapping'].pop(change_dict[
-                    'state_name'])
-            elif change_dict['cmd'] == exp_domain.CMD_RENAME_STATE:
-                prev_stats_dict['state_stats_mapping'][
-                    change_dict['new_state_name']] = (
-                        prev_stats_dict['state_stats_mapping'].pop(
-                            change_dict['old_state_name']))
+        for state_name in exp_versions_diff.states_added:
+            prev_stats_dict['state_stats_mapping'][state_name] = (
+                stats_domain.StateStats.create_default())
+
+        for state_name in exp_versions_diff.states_deleted:
```

Comment:
If a state gets renamed before being deleted, does states_deleted hold the new name or the old name of the state? If the former then this won't work.

(It should probably hold the old name if you want to apply the ExplorationVersionsDiff directly to the new object.)

Also would it make more sense for the rename-dict to be keyed by the old state name? That way the algorithm for transferring the data is simpler: delete stuff, then map old state names to new state names by iterating through the rename dict, then add new states.

Reply:
@seanlip Re the second point, does it make a difference? The data transfer is pretty simple right now as well.

Reply:
I think keying by old state names is simpler since you're transforming the old exploration into the new one. So you basically have OldExp and a function mapping stuff in OldExp to stuff in NewExp. Having the function going the other way round is conceptually a bit weird.